### ISSUE TITLE
Call for help: Windows Python Packagers

### ISSUE THREAD

**Issue by EmperorArthur** [mlox/mlox#36]
Hello all,

I have had the latest version of Mlox ready to go for quite a while now.  As in, over a year!  The only thing preventing a new release is that I don't know how to package a Python 3 application as an EXE on windows.  This is complicated because:
* Users are used to a single EXE, not a folder full of DLLs, etc...
* The current system relies on the program being run from within the Morrowind directory, and it relies on the Mlox directory being writable for chaching / portable operation.
* The updater for 'mlox_base' is actually incredibly powerful.  In fact, it's almost exploit level of powerful.  This gives us an option to in place update everyone running a copy of Mlox, or at least add a Version 2.  I don't want to silently update people, but need an updater channel to at least notify people of new versions.
* As is, simply using one of the self extracting EXEs that deletes the temp folder at the end has an issue where every time someone runs Mlox, they have to re-download 'mlox_base' which is not ideal.
* I'm using PyQt, which may or may not be default installed with Python 3 on Windows.

Fundamentally, it's just a question of having a way to tell users to update the program itself, and trying to find a balance between this things being a portable application or one that needs to be installed.  Especially since it needs to cache at least one file between runs, and users can add their own separate load order overrides as well.

### The good stuff:
As I said, the latest release is pretty much done.  The only question are a few things around the updater, and packaging the EXE for windows.  Here are the cool new features you can expect:

* Large re-write of the ordering system, and the underlying graph code used to determine load orders.
* Added a function, that's currently unavailable in the GUI, to generate actual graphviz graphs describing what's going re the load order.
* Moved to Python3.  Use this command to install the required dependencies `pip3 install PyQt5 libarchive-c appdirs`
* Moved to using PyQT instead of WxPython
  * We can now use graphical tools to design the GUI
  * GUI design and actual Python code are mostly separate.
  * We're now using an event based system to interact with the GUI.
  * Mostly it's just because I couldn't get WxPython to run with Python 3...

I know quite a few people use Mlox with OpenMW, and find it confusing to contribute to the project.  I'd like to make those things better, but just find it difficult when this major release has been stalled for so long.

**Comment by NullCascade** [comment]
We've had issues getting this built on Windows over at the Morrowind Modding Discord, but Greatness7 or I would probably be willing to help with releases.

Some of the issues we've run into:

* Encoding issues when iterating over files.
* Main window would not show.
* libarchive fails to import.

**Comment by EmperorArthur** [comment]
* Encoding issues when iterating over files.
* Main window would not show.

Both of those issues (at least when it comes to reading mlox_base.txt) have been fixed in git master.  Windows command line output also properly works now.

I have an [issue](https://github.com/pypa/pipenv/issues/3730) opened for libarchive on Pipenv.  Technically, I could make it an optional dependency, but that would require either downloading the uncompressed text file or bundling a separate utility.  Both of which I'm not super enthused about.

**Comment by EmperorArthur** [comment]
I have fixed the problems with packaging.  Those who wish to work on the software should use `pipenv`.  If you are interested in running in pure python mode use `pip3 install mlox`.

Otherwise there is now a windows exe that should just work.

@NullCascade I believe I have fixed the issues you encountered.
* I believe I have everything set up to always perform utf-8 decoding.
* Better resource handling should mean that the package can always find the GUI files.
* There are now multiple fallback options for extracting files, and the first option is a python only py7zr instead of libarchive.

Future work will revolve around code cleanup, or making the rules file more friendly.
